### PR TITLE
#9492: Change models matmul usage to ttnn

### LINKED ISSUE #9492: Move Matmul ops to ttnn
Consolidate matmuls in ttnn

### Plan
We propose to breakdown the work into smaller pieces:
- Replace usage of tt_eager based matmul ops in C++ with ttnn analogs ([example](https://github.com/tenstorrent/tt-metal/pull/9368/files#diff-3d9c8f1458a236487c85d72277c5e98ea5060cb889efae6e60e85baa124b9791))
- Replace usage in python Models, Demos, Examples, tests ([example](https://github.com/tenstorrent/tt-metal/pull/9368/files#diff-3decefc796146122ace7337ba1f3f717e90c19f1d24e4304dc690460fb77b3a1))
- Remove tt_lib bindings and entries from docs ([bindings](https://github.com/tenstorrent/tt-metal/pull/9368/files#diff-e8ae187c678c01424392f0efc63ac8c450686b4ccaf74aaed790941984e957e7), [docs](https://github.com/tenstorrent/tt-metal/pull/9368/files#diff-236cfcb211efba0ed290148fa450ca89f7b11dec8223b8cfb02afda350c9392d))
- Restructure `tt_eager/tt_dnn/op_library/bmm_op` and `ttnn/cpp/ttnn/operations/matmul.hpp` ([see new structure](https://github.com/tenstorrent/tt-metal/issues/9322))

#### Replacing usage in C++
Each usage of matmul should be replaced with a ttnn analog. For example `tt::operations::primary::matmul` --> `ttnn::matmul`.
Missing operations should be added to ttnn.
  - lots of usage in composite_ops.cpp, complex_ops.cpp, backward_ops.cpp 
  - lots of usage in tests, [example 1](https://github.com/tenstorrent/tt-metal/pull/9368/files#diff-f76070a3ff201ba6930e1fc34c698a486a3222a202e75841119661f5e7754ba4), [example 2](https://github.com/tenstorrent/tt-metal/pull/9368/files#diff-69c5eb0de84833ea5640c7da937d0145b3186a62d82cb3faaa6afae045a489daR123)

#### Replacing usage in Python
For every unary op, look for next entries in Tests/Sweeps, Demos, Models, Examples:
* `ttl.tensor.matmul`
* `tt_lib.tensor.matmul`
* `ttnn.primary.tensor.abs`
and replace them with `ttnn.matmul`. [Example](https://github.com/tenstorrent/tt-metal/pull/9368/files#diff-8a506d85107ff348dcb57916f47b780a4413e26a7e9094e66cea5a560da04fb4)

### Background
We got a lot of duplication.

We register unary ops in C++ in both tt_lib and ttnn:
- `ttnn/cpp/ttnn/operations/matmul.hpp` provides scaffold for unary ops and registers them in `ttnn`
- `tt_eager/tt_dnn/op_library/bmm/bmm_op.hpp` registers unary ops in tt_eager

We bind these ops in both:
- `ttnn/cpp/pybind11/operations/matmul.hpp`
- `tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_custom_bmm_ops.cpp`

We have tt_eager sweep tests, but they are not run on CI.

Duplication should go away. TTNN C++ and binding should remain and everything should get switched.

### Testing
For the best coverage, I recommend to run these workflows. If some of them fails, check if it is the same fail as on main:
-  Post commit CI
-  [post-commit] models tests
-  Device perf regressions and output report
-  Model perf regressions and output report
-  Nightly fast dispatch model regression CI
-  T3000 frequent tests
-  T3000 demo tests

### PR DESCRIPTION
**Pull request by bbradelTT**
### Ticket
Link to Github Issue https://github.com/tenstorrent/tt-metal/issues/9492

### Problem description
We're moving ops to ttnn

### What's changed
Make matmul ops in models point to ttnn matmul/linear

### Checklist
- [x] Post commit CI passes
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

https://github.com/tenstorrent/tt-metal/actions/runs/9681343061 All post-commits tests pass
https://github.com/tenstorrent/tt-metal/actions/runs/9682871493 Nightly fast dispatch tests fails with the same error as main
https://github.com/tenstorrent/tt-metal/actions/runs/9682879735/job/26717273487 (T3K) T3000 demo tests pass
https://github.com/tenstorrent/tt-metal/actions/runs/9683243284/job/26718212659 Device perf regressions and output report GS Device perf passes (the other one fails on main)
https://github.com/tenstorrent/tt-metal/actions/runs/9682876668 Model perf regressions and output report GS ones pass (WH is backed up)

2nd test run:

https://github.com/tenstorrent/tt-metal/actions/runs/9705332123 All post-commits tests pass
https://github.com/tenstorrent/tt-metal/actions/runs/9705034806 (T3K) T3000 model perf tests pass
https://github.com/tenstorrent/tt-metal/actions/runs/9704005481 (T3K) T3000 frequent tests pass
https://github.com/tenstorrent/tt-metal/actions/runs/9704008719 (T3K) T3000 demo tests pass except for mistral which this change doesn't touch and passed in an earlier attempt https://github.com/tenstorrent/tt-metal/actions/runs/9686832542

### CODE REVIEW

File: models/demos/metal_BERT_large_11/tt/bert_encoder.py [hunk 1]
```diff
@@ -142,13 +143,13 @@ def __init__(self, config, encoder_idx, state_dict, device, model_config, tt_cac
         if "OP7_SELFOUT_CONFIG" in model_config:
 
             def op7_mm_plus_bias(mha_res, attention_output_weight, attention_output_bias):
-                mha_out = tt_lib.operations.primary.matmul(
+                mha_out = ttnn.linear(
```
Comment (ayerofieiev-tt) [inline models/demos/metal_BERT_large_11/tt/bert_encoder.py:146]: can you educate me on the difference between `matmul` and `linear`?

Comment (ayerofieiev-tt) [reply models/demos/metal_BERT_large_11/tt/bert_encoder.py:146]: I see that in some cases we change to `linear` and sometimes on `matmul`

File: models/demos/t3000/llama2_70b/tests/perf/test_llama_matmul_perf.py [hunk 2]
```diff
@@ -91,12 +91,12 @@ def __init__(self, device):
 
     def __call__(self, x):
         # Assume interleaved input
-        ff_out = tt_lib.operations.primary.matmul_1d(
+        ff_out = ttnn.matmul(
```
Comment (ayerofieiev-tt) [inline models/demos/t3000/llama2_70b/tests/perf/test_llama_matmul_perf.py:94]: difference between 1d and general is hidden now?

File: models/experimental/llama2_70b/tests/unit_tests/test_qkv_matmul.py [hunk 3]
```diff
@@ -79,12 +79,12 @@ def forward(self, xs: tt_lib.tensor.Tensor) -> tt_lib.tensor.Tensor:
...
         for i in range(len(xs)):
             fused_query_key_value.append(
-                tt_lib.operations.primary.matmul(
+                ttnn.matmul(
                     xs[i],
                     self.qkv_list[i],
                     program_config=self.model_config["FUSED_QKV_MM_PROGCFG"],
-                    # output_mem_config=self.model_config["FUSED_QKV_MM_OUTPUT_MEMCFG"],
-                    # output_dtype=self.model_config["FUSED_QKV_MM_OUTPUT_DTYPE"],
+                    # memory_config=self.model_config["FUSED_QKV_MM_OUTPUT_MEMCFG"],
```
Comment (ayerofieiev-tt) [inline models/experimental/llama2_70b/tests/unit_tests/test_qkv_matmul.py:86]: lets remove?

File: models/demos/metal_BERT_large_11/tt/bert_encoder.py (hunk 1 above)
Comment (bbradelTT) [reply models/demos/metal_BERT_large_11/tt/bert_encoder.py:146]: They're the same op under the hood. In terms of the ttnn calls, the main differences are that linear 
- has two extra parameters: `bias` and `activation`
- does not propagate batching of b to the op (while matmul does)

File: models/demos/t3000/llama2_70b/tests/perf/test_llama_matmul_perf.py (hunk 2 above)
Comment (bbradelTT) [reply models/demos/t3000/llama2_70b/tests/perf/test_llama_matmul_perf.py:94]: Yes. And tt_lib.operations.primary.matmul_1d will be deleted.

File: models/experimental/llama2_70b/tt/llama_mlp_optimized.py [hunk 4]
```diff
@@ -117,22 +117,22 @@ def prefill_forward(self, x: List[tt_lib.tensor.Tensor]) -> List[tt_lib.tensor.T
         batch_dim = 1 if seq_len < max_mm_seq_len else seq_len // max_mm_seq_len  # Find the division factor
         x = ttnn.reshape(x, (1, batch_dim, seq_len // batch_dim, self.hidden_size))
 
-        w1_out = tt_lib.operations.primary.matmul(
+        w1_out = ttnn.matmul(
             x,
             self.w1,
             program_config=self.model_config["PADDED_FF1_MM_PROGCFG"],
-            # output_mem_config=self.model_config["MLP_BLOCK_SHARDED_MEMCFG"],
+            # memory_config=self.model_config["MLP_BLOCK_SHARDED_MEMCFG"],
```
Comment (ayerofieiev-tt) [inline models/experimental/llama2_70b/tt/llama_mlp_optimized.py:124]: lets remove?

File: models/experimental/llama2_70b/tt/llama_mlp_optimized.py [hunk 5]
```diff
@@ -117,22 +117,22 @@ def prefill_forward(self, x: List[tt_lib.tensor.Tensor]) -> List[tt_lib.tensor.T
...
+            dtype=self.model_config["BFLOAT16_DTYPE"],
         )
 
-        w3_out = tt_lib.operations.primary.matmul(
+        w3_out = ttnn.matmul(
             x,
             self.w3,
             program_config=self.model_config["PADDED_FF3_MM_PROGCFG"],
-            # output_mem_config=self.model_config["MLP_BLOCK_SHARDED_MEMCFG"],
+            # memory_config=self.model_config["MLP_BLOCK_SHARDED_MEMCFG"],
```
Comment (ayerofieiev-tt) [inline models/experimental/llama2_70b/tt/llama_mlp_optimized.py:133]: lets remove?

**Review by ayerofieiev-tt — APPROVED** [review]
🚀

**Review by boris-drazic — APPROVED** [review]

**Comment by tt-aho** [comment]
Can you also trigger the device perf pipeline?

**Comment by bbradelTT** [comment]
> Can you also trigger the device perf pipeline?

Ok. I'll do that and check that the failures are the same as on main.

File: models/demos/metal_BERT_large_11/tt/bert_encoder.py (hunk 1 above)
Comment (ayerofieiev-tt) [reply models/demos/metal_BERT_large_11/tt/bert_encoder.py:146]: Thank you!

**Review by eyonland — APPROVED** [review]

File: models/experimental/llama2_70b/tt/llama_mlp_optimized.py (hunk 4 above)
Comment (cglagovichTT) [reply models/experimental/llama2_70b/tt/llama_mlp_optimized.py:124]: yes, please remove

**Review by cglagovichTT — APPROVED** [review]
Approved on the condition that T3k frequent, T3k demo, and T3k perf pipelines are all green. Even if they are red today, we should only merge when this rebased branch is green.

File: models/experimental/llama2_70b/tests/unit_tests/test_qkv_matmul.py (hunk 3 above)
Comment (bbradelTT) [reply models/experimental/llama2_70b/tests/unit_tests/test_qkv_matmul.py]: Done

File: models/experimental/llama2_70b/tt/llama_mlp_optimized.py (hunk 4 above)
Comment (bbradelTT) [reply models/experimental/llama2_70b/tt/llama_mlp_optimized.py]: Done

File: models/experimental/llama2_70b/tt/llama_mlp_optimized.py (hunk 5 above)
Comment (bbradelTT) [reply models/experimental/llama2_70b/tt/llama_mlp_optimized.py]: Done

**Comment by bbradelTT** [comment]
> Approved on the condition that T3k frequent, T3k demo, and T3k perf pipelines are all green. Even if they are red today, we should only merge when this rebased branch is green.

T3k perf hasn't been green for two weeks. Is there any plan about when these pipelines will be green?

**Comment by cglagovichTT** [comment]
> > Approved on the condition that T3k frequent, T3k demo, and T3k perf pipelines are all green. Even if they are red today, we should only merge when this rebased branch is green.
> 
> T3k perf hasn't been green for two weeks. Is there any plan about when these pipelines will be green?

We will merge a PR this afternoon which temporarily lowers the perf threshold to make it green.

**Comment by bbradelTT** [comment]
> > > Approved on the condition that T3k frequent, T3k demo, and T3k perf pipelines are all green. Even if they are red today, we should only merge when this rebased branch is green.
> > 
> > 
> > T3k perf hasn't been green for two weeks. Is there any plan about when these pipelines will be green?
> 
> We will merge a PR this afternoon which temporarily lowers the perf threshold to make it green.

Cool. While I'm asking, what about the timeout failures in [T3K] T3000 frequent tests, e.g. https://github.com/tenstorrent/tt-metal/actions/runs/9681494756 ?

**Review by TT-BrianLiu — APPROVED** [review]
Looks good to me. One question: 
- `ttnn.matmul` has a `use_1d_systolic` flag to force our semi-automatic program logic to select matmul 1D. For `matmul_1d`, should we be setting this flag or let the semi-automatic logic infer between 2D vs 1D?

**Review by tt-aho — APPROVED** [review]

**Comment by bbradelTT** [comment]
> Looks good to me. One question:
> 
> * `ttnn.matmul` has a `use_1d_systolic` flag to force our semi-automatic program logic to select matmul 1D. For `matmul_1d`, should we be setting this flag or let the semi-automatic logic infer between 2D vs 1D?

I would prefer to let the semi-automatic logic infer between 2D vs 1D.

That flag should probably be deprecated. The functionality to force 1D was removed. That flag is only used to pass in a core grid which triggers the logic that was there in python to make existing tests pass.

File: models/demos/resnet/tt/metalResnetBlock50.py [hunk 6]
```diff
@@ -123,18 +123,18 @@ def ResnetLinear(
     if matmul_config is None:
 
         def linear_(act):
-            return tt_lib.tensor.resnet_matmul(act, weight_T, bias, output_mem_config)
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/demos/resnet/tt/metalResnetBlock50.py:126]: Yes, resnet_matmul will be deleted in a separate pr.

**Pull request merged**